include/libhal-sensor/multi/bmp180.hpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2024 Khalil Estell
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+
+#include <libhal/i2c.hpp>
+#include <libhal/units.hpp>
+
+namespace hal::sensor {
+
+/**
+ * @brief Driver for the BMP180 barometric pressure/temperature sensor
+ */
+
+class bmp180
+{
+public:
+  /**
+   * @brief Oversampling rate defined with the mode names as well as sample
+   * fetch times.
+   */
+  enum class oversampling_rate : hal::byte
+  {
+    /// 3 internal samples by sensor
+    ulta_low_power_mode_4500us = 0x00,
+    /// 5 internal samples by sensor
+    standard_mode_7500us = 0x01,
+    /// 9 internal samples by sensor
+    high_resolution_mode_13500us = 0x02,
+    /// 17 internal samples by sensor
+    ultra_high_resolution_mode_25500us = 0x03,
+  };
+
+  /**
+   * @brief Stores the temperature and pressure data that are both found when
+   * doing the pressure()
+   */
+  struct pressure_results
+  {
+    hal::celsius temperature;
+    float pressure;
+  };
+
+  /**
+   * @brief Construct a bmp180 driver
+   *
+   * @param p_i2c - The driver for the SPI bus the BMP180 is connected to.
+   * @param p_oversampling_setting - The specified oversampling setting which
+   * sets the amount of internal samples the sensor takes at each request. This
+   * cannot be changed after the object is constructed.
+   * @throws hal::no_such_device - when an invalid BMP180 device is detected.
+   * BMP180 devices have a read-only ID register which allows a microcontroller
+   * to determine what device it is connected to. This register will be read and
+   * if it does not match the expected value, this exception is thrown.
+   * @throws hal::io_error - when something is wrong with the data
+   * communication. The calibration data that the sensor uses need to all have
+   * non-boundary values, if any are 0x0000 or 0xFFFF then this is thrown.
+   */
+  explicit bmp180(hal::i2c& p_i2c,
+                  oversampling_rate p_oversampling_setting =
+                    oversampling_rate::ulta_low_power_mode_4500us);
+
```

Comment:
```suggestion
  bmp180(hal::i2c& p_i2c,
                  oversampling_rate p_oversampling_setting =
                    oversampling_rate::ulta_low_power_mode_4500us);

```

Explicit is only needed when there is a single parameter for a class. C++ allows the assignment operator to act as a constructor if there is just one element which results in a sort of implicit conversion.

Reply:
Since the second parameter p_oversampling_setting has a default value, then the bmp180 constructor can be called (and treated) as just having the single parameter p_i2c. In this scenario, testing revealed that the assignment operator can be used to declare a bmp180 object. As a result, it's necessary to retain the explicit specifier.